programs/drift/src/state/spot_fulfillment_params.rs

```diff
@@ -0,0 +1,394 @@
+use crate::controller::serum::{invoke_new_order, invoke_settle_funds};
+use crate::error::{DriftResult, ErrorCode};
+use crate::math::safe_math::SafeMath;
+use crate::math::safe_unwrap::SafeUnwrap;
+use crate::math::serum::{
+    calculate_serum_limit_price, calculate_serum_max_coin_qty,
+    calculate_serum_max_native_pc_quantity,
+};
+use crate::math::spot_withdraw::validate_spot_market_vault_amount;
+use crate::state::events::OrderActionExplanation;
+use crate::state::serum::{get_best_bid_and_ask, load_open_orders, load_serum_market};
+use crate::state::spot_market::{SpotBalanceType, SpotMarket};
+use crate::PositionDirection;
+use anchor_lang::prelude::{Account, Program};
+use anchor_lang::ToAccountInfo;
+use anchor_spl::token::{Token, TokenAccount};
+use serum_dex::instruction::{NewOrderInstructionV3, SelfTradeBehavior};
+use serum_dex::matching::Side;
+use solana_program::account_info::AccountInfo;
+use solana_program::msg;
+use std::cell::Ref;
+use std::num::NonZeroU64;
+
+pub trait SpotFulfillmentParams {
+    fn is_external(&self) -> bool;
+
+    fn get_best_bid_ask(
+        &self,
+        base_market_decimals: u32,
+    ) -> DriftResult<(Option<u64>, Option<u64>)>;
+
+    fn fulfill_order(
+        &mut self,
+        taker_direction: PositionDirection,
+        taker_price: u64,
+        taker_base_asset_amount: u64,
+        taker_max_quote_asset_amount: u64,
+        base_market_decimals: u32,
+        now: i64,
+    ) -> DriftResult<ExternalSpotFill>;
+
+    fn get_order_action_explanation(&self) -> DriftResult<OrderActionExplanation>;
+
+    fn validate_vault_amounts(
+        &self,
+        base_market: &Ref<SpotMarket>,
+        quote_market: &Ref<SpotMarket>,
+    ) -> DriftResult<()>;
+}
+
+pub struct MatchFulfillmentParams<'a> {
```

Comment:
nit: Feels cleaner to separate all of the impls into their own files in a subdirectory